From 39c9a91bc0920db3a4acc38ddf84cb94dcafb927 Mon Sep 17 00:00:00 2001
From: Daniel Golle <daniel@makrotopia.org>
Date: Thu, 16 Mar 2023 16:31:51 +0530
Subject: [PATCH 044/500] OpenWrt:
 492-try-auto-mounting-ubi0-rootfs-in-init-do_mounts.c.patch

try auto-mounting ubi0:rootfs in init/do_mounts.c

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Change-Id: I64d5e42242c75edc7b2446c25f008956a9f3f26d
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
(cherry picked from commit 6eac4615d18966dbc6b5796bafd9ad5777f336f3)
---
 init/do_mounts.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/init/do_mounts.c b/init/do_mounts.c
index 5dfd30b13f48..8e5d30ddb79b 100644
--- a/init/do_mounts.c
+++ b/init/do_mounts.c
@@ -383,6 +383,29 @@ static inline void mount_block_root(char *root_device_name)
 }
 #endif /* CONFIG_BLOCK */
 
+#ifdef CONFIG_MTD_ROOTFS_ROOT_DEV
+static int __init mount_ubi_rootfs(void)
+{
+	int flags = MS_SILENT;
+	int err, tried = 0;
+
+	while (tried < 2) {
+		err = do_mount_root("ubi0:rootfs", "ubifs", flags, \
+					root_mount_data);
+		switch (err) {
+			case -EACCES:
+				flags |= MS_RDONLY;
+				tried++;
+				break;
+			default:
+				return err;
+		}
+	}
+
+	return -EINVAL;
+}
+#endif
+
 void __init mount_root(char *root_device_name)
 {
 	switch (ROOT_DEV) {
@@ -397,6 +420,11 @@ void __init mount_root(char *root_device_name)
 				   root_mountflags);
 		break;
 	case 0:
+
+#ifdef CONFIG_MTD_ROOTFS_ROOT_DEV
+	if (!mount_ubi_rootfs())
+		return;
+#endif
 		if (root_device_name && root_fs_names &&
 		    mount_nodev_root(root_device_name) == 0)
 			break;
-- 
2.34.1

